lib/groups.js

```diff
@@ -0,0 +1,144 @@
+    if (join_code.length !== 4) {
+      throw new Error('invalid length of join code');
+    }
+    let params = [
+      res.locals.assessment.id,
+      res.locals.user.user_id,
+      res.locals.authn_user.user_id,
+      group_name,
+      join_code,
+    ];
+    sqldb.call('group_users_insert', params, function (err, _result) {
+      if (err) {
+        var joinErr = err;
+        let params = {
+          assessment_id: res.locals.assessment.id,
+        };
+        sqldb.query(sql.get_config_info, params, function (err, result) {
+          if (ERR(err, callback)) return;
+          res.locals.permissions = result.rows[0];
+          res.locals.groupsize = 0;
+          res.locals.used_join_code = joinCode;
+          //display the error on frontend
+          callback(null, joinErr);
+          return;
+        });
+      } else {
+        callback();
+      }
+    });
+  } else {
+    // the join code input by user is not valid (not in format of groupname+4-character)
+    let params = {
+      assessment_id: res.locals.assessment.id,
+    };
+    sqldb.query(sql.get_config_info, params, function (err, result) {
+      if (ERR(err, callback)) return;
+      res.locals.permissions = result.rows[0];
+      res.locals.groupsize = 0;
+      res.locals.used_join_code = joinCode;
+      //display the error on frontend
+      callback(null, !validJoinCode);
+    });
+  }
+};
+
+module.exports.createGroup = function (groupName, res, callback) {
+  const params = {
+    assessment_id: res.locals.assessment.id,
+    user_id: res.locals.user.user_id,
+    authn_user_id: res.locals.authn_user.user_id,
+    group_name: groupName,
+  };
+  //alpha and numeric characters only
+  let invalidGroupName = true;
+  const letters = /^[0-9a-zA-Z]+$/;
+  if (groupName.length <= 30 && groupName.match(letters)) {
+    invalidGroupName = false;
+    //try to create a group
+    sqldb.query(sql.create_group, params, function (err, _result) {
+      if (!err) {
+        callback();
+      } else {
+        let createErr = err;
+        sqldb.query(sql.get_config_info, params, function (err, result) {
+          if (ERR(err, callback)) return;
+          res.locals.permissions = result.rows[0];
+          res.locals.groupsize = 0;
+          res.locals.uniqueGroupName = true;
+          callback(null, createErr);
+        });
+      }
+    });
+  }
+  if (invalidGroupName) {
+    sqldb.query(sql.get_config_info, params, function (err, result) {
+      if (ERR(err, callback)) return;
+      res.locals.permissions = result.rows[0];
+      res.locals.groupsize = 0;
+      res.locals.invalidGroupName = true;
+      callback(null, invalidGroupName);
+    });
+  }
+};
+module.exports.leaveGroup = function (res, callback) {
+  const params = {
+    assessment_id: res.locals.assessment.id,
+    user_id: res.locals.user.user_id,
+    authn_user_id: res.locals.authn_user.user_id,
+  };
+  sqldb.query(sql.leave_group, params, function (err, _result) {
```

Comment:
If the result is ignored, you can just omit the parameter.
```suggestion
  sqldb.query(sql.leave_group, params, function (err) {
```

Reply:
Got it! Fixed in [2e62979](https://github.com/PrairieLearn/PrairieLearn/pull/5290/commits/2e62979dea6d192225699af5a53d59d1946cb496)